src/webgpu/api/validation/encoding/encoder_state.spec.ts

```diff
@@ -19,4 +19,67 @@ TODO:
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
 import { ValidationTest } from '../validation_test.js';
 
-export const g = makeTestGroup(ValidationTest);
+class F extends ValidationTest {
+  beginRenderPass(commandEncoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const attachmentTexture = this.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 16, height: 16, depthOrArrayLayers: 1 },
+      usage: GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    return commandEncoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: attachmentTexture.createView(),
+          clearValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+    });
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('pass_end_invalid_order')
+  .desc(
+    `
+  Test that beginning a [compute|render] pass before ending the previous pass causes an error.
+  `
+  )
+  .paramsSubcasesOnly(u =>
+    u
+      .combine(`IsRenderPassFirst`, [true, false])
+      .combine(`IsComputePassSecond`, [true, false])
+      .combine('endPasses', [[], [0], [1], [0, 1], [1, 0]])
+  )
+  .fn(async t => {
+    const { IsRenderPassFirst, IsComputePassSecond, endPasses } = t.params;
+
+    const encoder = t.device.createCommandEncoder();
+
+    let firstPass;
+    if (IsRenderPassFirst) {
+      firstPass = t.beginRenderPass(encoder);
+    } else {
+      firstPass = encoder.beginComputePass();
+    }
+
+    // Begin a second pass before ending the previous pass.
+    let secondPass;
+    if (IsComputePassSecond) {
+      secondPass = encoder.beginComputePass();
+    } else {
+      secondPass = t.beginRenderPass(encoder);
+    }
+
+    const passes = [firstPass, secondPass];
+    for (const index of endPasses) {
+      passes[index].end();
+    }
+
+    t.expectValidationError(() => {
+      encoder.finish();
+    }, true);
```

Comment:
Maybe we should consider trying to have a control case here. Without a control case it's hard to be sure there isn't something wrong with the test other than what we're trying to test. However... I'm not really sure how to do it. It's probably fine if we can't think of a good way to add a control case.

Reply:
yes, it was hard to add a control case from my side. That's why I just checked a validation error happens. Then, let me add a TODO for the control case. And, let's update this test if we have an idea to add a control case next time.